wagtail/contrib/search_promotions/views.py

```diff
@@ -125,121 +125,113 @@ def save_searchpicks(query, new_query, searchpicks_formset):
-                for search_pick in searchpicks_formset.new_objects:
-                    log(search_pick, "wagtail.create")
-                messages.success(
-                    request,
-                    _("Editor's picks for '%(query)s' created.") % {"query": query},
-                    buttons=[
-                        messages.button(
-                            reverse("wagtailsearchpromotions:edit", args=(query.id,)),
-                            _("Edit"),
-                        )
-                    ],
-                )
-                return redirect("wagtailsearchpromotions:index")
-            else:
-                if len(searchpicks_formset.non_form_errors()):
-                    # formset level error (e.g. no forms submitted)
-                    messages.error(
-                        request,
-                        " ".join(
-                            error for error in searchpicks_formset.non_form_errors()
-                        ),
-                    )
-                else:
-                    # specific errors will be displayed within form fields
-                    messages.error(
-                        request,
-                        _("Recommendations have not been created due to errors"),
-                    )
         else:
-            searchpicks_formset = forms.SearchPromotionsFormSet()
-    else:
-        query_form = forms.QueryForm()
-        searchpicks_formset = forms.SearchPromotionsFormSet()
+            context["searchpicks_formset"] = forms.SearchPromotionsFormSet()
 
-    return TemplateResponse(
-        request,
-        "wagtailsearchpromotions/add.html",
-        {
-            "query_form": query_form,
-            "searchpicks_formset": searchpicks_formset,
-            "form_media": query_form.media + searchpicks_formset.media,
-        },
-    )
+        context["form_media"] = (
+            self.get_form().media + context["searchpicks_formset"].media
+        )
+        return context
 
+    def form_valid(self, form):
+        query = form.save()
 
-@permission_required("wagtailsearchpromotions.change_searchpromotion")
-def edit(request, query_id):
-    query = get_object_or_404(Query, id=query_id)
+        searchpicks_formset = forms.SearchPromotionsFormSet(
+            self.request.POST, instance=query
+        )
+
+        if searchpicks_formset.is_valid():
+            save_searchpicks(query, query, searchpicks_formset)
+
+            for search_pick in searchpicks_formset.new_objects:
+                log(search_pick, "wagtail.create")
+
+            self.success_message = self.success_message % {"query": query}
+            self.success_url = reverse(self.index_url_name)
+            return super().form_valid(form)
+        else:
+            self.error_message = " ".join(
+                error for error in searchpicks_formset.non_form_errors()
+            )
+            return self.form_invalid(form)
+
+
+class EditSearchPromotion(generic.EditView):
```

Comment:
This is also missing a `permission_policy`, and would be nice if we can enable breadcrumbs too